pkg/workload/workload.go

```diff
@@ -175,13 +186,103 @@ func Tuples(count int, fn func(int) []interface{}) BatchedTuples {
 		NumTotal:   count,
 	}
 	if fn != nil {
-		t.Batch = func(batchIdx int) [][]interface{} {
-			return [][]interface{}{fn(batchIdx)}
+		t.FillBatch = func(batchIdx int, cb coldata.Batch, _ *bufalloc.ByteAllocator) {
+			row := fn(batchIdx)
+			// It'd be nice if we could share this between calls to save the work and
+			// the allocation.
+			colTypes := make([]types.T, len(row))
+			for i, datum := range row {
+				if datum == nil {
+					// WIP what do we do here
+					colTypes[i] = types.Bytes
+				} else {
+					switch datum.(type) {
+					case time.Time:
+						colTypes[i] = types.Bytes
+					default:
+						colTypes[i] = types.FromGoType(datum)
+					}
+				}
+			}
+			cb.Reset(colTypes, 1)
+			for colIdx, col := range cb.ColVecs() {
+				switch d := row[colIdx].(type) {
+				case nil:
+					col.SetNull(0)
+				case bool:
+					col.Bool()[0] = d
+				case int:
+					col.Int64()[0] = int64(d)
+				case float64:
+					col.Float64()[0] = d
+				case string:
+					col.Bytes()[0] = []byte(d)
+				case []byte:
+					col.Bytes()[0] = d
+				case time.Time:
+					col.Bytes()[0] = []byte(d.Round(time.Microsecond).UTC().Format(tree.TimestampOutputFormat))
+				default:
+					panic(fmt.Sprintf(`unhandled datum type %T`, d))
+				}
+			}
 		}
 	}
 	return t
 }
 
+// BatchRows is a function to deterministically compute a row-batch of tuples
+// given its index. BatchRows doesn't attempt any reuse and so is allocation
+// heavy. In performance-critical code, FillBatch should be used directly,
+// instead.
+func (b BatchedTuples) BatchRows(batchIdx int) [][]interface{} {
+	cb := coldata.NewMemBatch(nil)
+	var a bufalloc.ByteAllocator
+	b.FillBatch(batchIdx, cb, &a)
+	return ColBatchToRows(cb)
+}
+
+// ColBatchToRows materializes the columnar data in a coldata.Batch into rows.
+func ColBatchToRows(cb coldata.Batch) [][]interface{} {
+	numRows, numCols := int(cb.Length()), cb.Width()
+	// Allocate all the []interface{} row slices in one go.
+	datums := make([]interface{}, numRows*numCols)
+	for colIdx, col := range cb.ColVecs() {
+		switch col.Type() {
```

Comment:
Like, here, you should really have the type information passed in, rather than having to check the ColVec, in my opinion. `ColBatchToRows` should also take a `[]types.T`.